BufferGeometry: Implemented morphTargetsRelative

zeux opened the pull request:
This change implements an addition to core BufferGeometry class, morphTargetsRelative.

By default, morphTargetsRelative is false which means that the morph targets contain absolute position/normal data. Setting this boolean to true forces three.js to interpret the target data as deltas from the base.

To implement rendering support, instead of duplicating shader code, the base mesh gets a weight that is set to 1 for relative targets and 1 - sum(weights) for absolute targets. The math ends up matching the original math with this:

base + w1 * (target1 - base) + w2 * (target2 - base) = base - w1 * base - w2 * base + w1 * target1 + w2 * target2 = base * (1 - w1 - w2) + w1 * target1 + w2 * target2

In JavaScript code we need to occasionally check the morphTargetsRelative value to determine the correct action.

As a result, glTF loader (in theory FBX loader as well, but I'm not an expert in FBX loader code so I didn't want to change this in this PR) doesn't need to duplicate morph target data. This fixes some issues with morph target format encodings, saves memory in some cases, and removes unnecessary data copying from the loading step.

I've tested this change on built-in examples that exercise morph target support as well as some custom glTF files processed by gltfpack. Also note that the change is backwards-compatible in that using an older version of GLTFLoader.js works as expected (by using the absolute morph target mode).

Fixes #17608.

zeux committed:
Implement basic support for morphTargetsRelative

This adds the attribute and the necessary plumbing for various JS
functions; there's no support for specifying this for the shader to
consume yet.

zeux committed:
Implement support for morphTargetsRelative in WebGL

The shader code is now a bit simpler and relies on the precomputed base
influence value; JS code sets it up appropriately.

zeux committed:
Use morphTargetsRelative in GLTF loader

This uses the newly added functionality for loading morph target data in
GLTF loader; there's no need to clone attribute data anymore.

zeux committed:
Remove clamp-to-zero from WebGL morph target code

On further reflection it shouldn't be necessary - it's probably better
to maintain "exact" match between the shader code and CPU-side code that
may use (pos - morph) * w directly.

zeux committed:
Update GLTF exporter and BufferGeometryUtils to support relative morphs

The code in GLTF exporter is probably suboptimal - it's not clear that
we need to clone the attribute if it's already relative.

zeux committed:
Fix unit test by adjusting the gold data

looeee commented:
This would make implementation of  morphtargets in the `FBXLoader` much simpler.

donmccurdy commented:
I'd like to do some performance profiling to quantify the difference, but my hunch is that this will be a significant parse time improvement for glTF models relying on morph targets, such as https://mrdoob.github.io/rome-gltf/.

takahirox commented:
Nice update. I'm sure this change simplifies the glTF loader/exporter and reduced memory consumption.

I haven't looked into the implementation in the detail yet but I guess serialization/deserialization (`.toJSON()` and `BufferGeometryLoader`) should update to support `morphTargetsRelative`?

zeux commented:
I missed the BufferGeometryLoader somehow - will fix!

zeux committed:
Implement relative morph target support in BufferGeometryLoader

Also fix toNonIndexed to copy the attribute

zeux committed:
Implement support for relativeMorphTargets in Projector code

zeux committed:
Fix computeBoundingSphere for relative morph targets

donmccurdy commented on src/core/BufferGeometry.js:
```diff
@@ -645,8 +655,17 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
+					if ( this.morphTargetsRelative ) {
+
+						_box.boundingBox.expandByVector( _boxMorphTargets.min );
+						_box.boundingBox.expandByVector( _boxMorphTargets.max );
```
Should be `_box.expandByVector(` here I think.

zeux replied:
Yup - just noticed this as well when testing decals. expandByVector also doesn't do quite what I thought it's doing so I think this is slightly incorrect - I'll double check and fix this.

donmccurdy commented:
Some performance numbers for context. Testing against the glTF files in [this folder](https://github.com/mrdoob/rome-gltf/tree/master/files/models/black_soup), all of which contain morph targets, I found a couple interesting things.

First, parsing time was actually dominated by the embedded Data URIs in the `.gltf` files. I'd realized those were wasting file size, but hadn't realized they affected parsing time so much. So the takeway there is that it's better to either use `.gltf` with separate `.bin` files, or just use embedded binary `.glb` files. And usually `.glb` would be the better choice, to save network roundtrips.

With the slower `.gltf` files, rewriting morph targets represented 13% of parse time. With the faster `.glb` files, it was about 34%. In either case, using relative morph targets cut that to pretty much 0%. I don't necessarily know that these are representative files, and it was surprising to see how much variance there was in the individual runs. But the models all do make heavy use of morph targets.

A couple charts, showing the 7 models tested 3x each. Time spent processing morph targets is indicated in red.

<img width="1227" alt="gltf_parse" src="https://user-images.githubusercontent.com/1848368/66180981-da2ae980-e623-11e9-8c93-1c880dfad1ad.png">

<img width="1227" alt="glb_parse" src="https://user-images.githubusercontent.com/1848368/66180985-de570700-e623-11e9-8e4d-7545b949505d.png">

The important chart here is the one in the bottom left – I'd take that as a pretty good indicator that this makes a measurable difference on parsing time of real models.

zeux committed:
Fix computeBoundingBox/Sphere for relative morph targets

expandByVector is not correct to call here; if a morph target has a min
of [1,1,1], it should *not* change the box.min but it should expand
box.max by 1.

To avoid doing per-component math we use addVectors to get the corners
of the box, and use expandByPoint to union the box with the resulting
point.

zeux commented:
I've updated the change to fix a few stray examples / code that slipped my initial notice, and also fix the computeBoundingBox/Sphere functions to use precise (I think) formulas.

@donmccurdy Thanks for gathering the numbers!

donmccurdy commented:
@mrdoob per https://github.com/mrdoob/three.js/pull/17649#issuecomment-538226384, this would make GLTFLoader parsing more efficient. What do you think about (a) whether to add the property, and (b) how it should be located and named?

zeux commented:
FWIW while I'm not attached to the naming I think the property should be on BufferGeometry - this is because otherwise the data stored in BufferGeometry.morphAttributes becomes impossible to interpret outside of context. For example, if the property lived on the Material, changing materials would break the morphed shape visually if the value isn't fixed up. With this interface, BufferGeometry is "portable" - it can be migrated between objects / meshes / materials without losing the semantics.

zeux commented:
Oh and also given that both glTF and FBX loaders use relative data it seems clear that relative data is an important use case. Some formats like COLLADA support both - interestingly, if you look at COLLADA spec, they use the exact same formulation as this change:

![image](https://user-images.githubusercontent.com/1106629/66486279-28464f80-ea5f-11e9-82f4-36a6403825c5.png)

zeux commented:
I've rebased this against dev because of a trivial conflict with BufferGeometry.d.ts; should be cleanly mergeable again. @mrdoob please let me know if there's anything else to be done here, as per above this seems beneficial and complete. I could rename morphTargetsRelative to something like morphRelative if that seems more appropriate.

Mugen87 commented on src/core/BufferGeometry.js:
```diff
@@ -37,6 +37,7 @@ function BufferGeometry() {
 	this.attributes = {};
 
 	this.morphAttributes = {};
+	this.morphTargetsRelative = false;
```
Can you please add this new property to the documentation (https://threejs.org/docs/index.html#api/en/core/BufferGeometry)?

zeux replied:
Done! I'm assuming the Chinese translation is done separately; the best I can do is Google Translate 😅

Mugen87 commented on src/renderers/shaders/ShaderChunk/morphtarget_vertex.glsl.js:
```diff
@@ -1,17 +1,18 @@
-	transformed += ( morphTarget1 - position ) * morphTargetInfluences[ 1 ];
-	transformed += ( morphTarget2 - position ) * morphTargetInfluences[ 2 ];
-	transformed += ( morphTarget3 - position ) * morphTargetInfluences[ 3 ];
+	transformed *= morphTargetBaseInfluence;
```
Any chances to add a link to the formula of this code? Or maybe to a reference that explains what's going on in the next lines? I guess that would make the code more clear and accessible for the interested reader.

zeux replied:
I've updated both shaders (position / normal) and JS-side code to explain this a bit better.

Mugen87 commented:
I've added two minor comments. Apart from that, the change looks good!

zeux committed:
Add morphTargetsRelative to documentation

zeux committed:
Add GLSL comments to explain the math a bit

zeux committed:
Add comments to WebGLMorphtargets.js as well

Mugen87 commented on docs/api/en/core/BufferGeometry.html:
```diff
@@ -144,6 +144,14 @@ <h3>[property:Object morphAttributes]</h3>
+			Used to control the morph target behavior; when set to true, the morph target data is treated as relative changes to the base geometry,
+			instead of completely replacing the base geometry data.
+
+			Default is *true*.
```
Minor: The default is `false`.

zeux committed:
Update BufferGeometry.html

Fix default value in documentation

WestLangley commented on docs/api/en/core/BufferGeometry.html:
```diff
@@ -144,6 +144,14 @@ <h3>[property:Object morphAttributes]</h3>
+		<h3>[property:Boolean morphTargetsRelative]</h3>
+		<p>
+			Used to control the morph target behavior; when set to true, the morph target data is treated as relative changes to the base geometry,
+			instead of completely replacing the base geometry data.
```
When set to true, the morph target data is treated as relative offsets, rather than as absolute positions.

zeux replied:
Thanks! This wording is much better; I've adjusted this to say "positions/normals".

WestLangley commented on examples/js/exporters/GLTFExporter.js:
```diff
@@ -1339,14 +1339,18 @@ THREE.GLTFExporter.prototype = {
-								attribute.getY( j ) - baseAttribute.getY( j ),
-								attribute.getZ( j ) - baseAttribute.getZ( j )
-							);
+						if ( !geometry.morphTargetsRelative ) {
```
nit: formatting
```js
 ! geometry.morphTargetsRelative
```

zeux committed:
Improve wording in the documentation

zeux committed:
Fix formatting to make sure mrdoob approves.

donmccurdy reviewed (approved):
I'd support making a change like this to allow relative morph targets, especially given https://github.com/mrdoob/three.js/issues/17608#issuecomment-538228742. I'm open to other APIs for it, but this PR seems like the most direct approach from where the API is today.

zeux commented:
@mrdoob please let me know if any further changes are required

mrdoob commented:
@zeux 

Sorry for the delay.

In your experience, do game engines usually support both modes? Or is it usually only relative mode? (Tempted of only supporting relative mode and have a method to convert absolute to relative to help with breakage)

Also, does it only apply to vertex positions? Or also to normals and colors somehow?

takahirox commented:
@mrdoob https://github.com/mrdoob/three.js/issues/17608#issuecomment-538228742 may help

zeux commented:
Yeah so resummarizing - glTF and FBX work with relative data; COLLADA allows both; keyframed animations from MD2 use absolute data.

I think there are two use cases for morph targets - one prevalent one, for blend shapes - morphing between multiple variants of the same mesh where usually only part of the mesh gets deformed and it’s common to apply many shapes with fractional coefficients - think facial expressions (where you always overlay corrections on top of base pose). And one less common one, for morphing between substantially different poses - e.g. if you want to morph between meshes with unrelated structure relative data doesn’t make as much sense.

My overall feeling is that relative targets are the most important, and absolute targets could be either not supported or emulated through absolute targets. However, this is a large breaking change.

So I would recommend doing this gradually:

- Adding support for relative targets (this PR)
- Changing loaders that work with relative data to use this support directly (this PR does it for GLTF but there’s also FBX and COLLADA)
- Surveying users about whether their use cases need absolute support
- If it’s considered unnecessary to support absolute mode, migrating loaders that expect absolute mode (MD2, possibly COLLADA when it uses absolute data) to transform absolute to relative, and removing support separately.

The cost of supporting both isn’t very large so I would err on the side of caution, but I don’t know the philosophy of Three.JS wrt breaking changes.

zeux commented:
Oh and the morphs are usually relative or absolute for all attributes, including normals.

The general form here really is a full weighted sum of all targets, with a custom weight for base target. If the fact that there are many places that have to check the relative state is a concern, I can refactor this to use the general weighting math from the shader on JavaScript side? That would remove most of the conditions. I would prefer to do it in a separate PR but I can do it as part of this one if necessary.

mrdoob commented:
Thanks for summarising!

I think I vote for making this a breaking change: support only relative morphtargets and adding a method that converts absolute morphattributes to relative.

zeux commented:
Do you mean to do this in one step? This also means that the three.js release with this change will silently break components such as GLTFLoader if they aren't using the version from this change which seems unsatisfying. The original intent for this PR was to make a change that is safe to deploy, so that deprecation can be done separately.

Mugen87 commented:
>This also means that the three.js release with this change will silently break components such as GLTFLoader if they aren't using the version from this change which seems unsatisfying.

We normally do not support this scenario anyway. Check out the following notes from the [docs](https://threejs.org/docs/index.html#manual/en/introduction/Import-via-modules) (last section):

> Note: When using code from the examples directory, it's important that all files match the version of your three.js main file. For example, it's not acceptable to use GLTFLoader and OrbitControls from R96 together with three.js R103.

mrdoob commented:
> Do you mean to do this in one step?

Yeah

> The original intent for this PR was to make a change that is safe to deploy, so that deprecation can be done separately.

Yes, but it also introduces new API (`morphTargetsRelative`) which we'll have to serialise, deserialise, maintain, and eventually add to the legacy file.

Seems easier to have a breaking change and point people to a method that fixes it.

mrdoob commented:
We can do it in two steps actually.
I'll merge this and I'll do another PR with the second step so we can revert just in case.

zeux commented:
Sorry I’ve been short on time the last few days. Two steps sound good - that’s my preference anyway. It is my understanding that JSON export isn’t a long term compatibility target either? I can remove the extra attribute from JSON export from this change so that we don’t have intermediate changes to the format.

mrdoob commented:
Better to merge it at the beginning of `r111`.

mrdoob commented:
Thanks!

mrdoob commented:
> This would make implementation of  morphtargets in the `FBXLoader` simpler.

@looeee Any chance`FBXLoader` can be updated to use `morphTargetsRelative`?

donmccurdy commented:
Thank you @zeux! :)

looeee commented:
@mrdoob Are you still considering [removing the `FBXLoader` at some point](https://github.com/mrdoob/three.js/issues/15745#issuecomment-526861987)? 

If so I'm not sure that it's worth spending the time to update it.

mrdoob commented:
@looeee well... removing code from it will help make it less intimidating and maybe that way we'll get more eyes on it.

looeee commented:
> maybe that way we'll get more eyes on it.

Well, the code is now about 3 lines shorter so let's see how that goes 😁 

 #17890
